cpp/src/kpca/kpca.cuh

```diff
@@ -0,0 +1,180 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <thrust/fill.h>
+#include <thrust/device_ptr.h>
+#include <cuml/matrix/kernelparams.h>
+#include <raft/linalg/cublas_wrappers.h>
+#include <cuml/common/device_buffer.hpp>
+#include <cuml/decomposition/params.hpp>
+#include <matrix/kernelfactory.cuh>
+#include <matrix/kernelmatrices.cuh>
+#include <raft/cuda_utils.cuh>
+#include <raft/handle.hpp>
+#include <raft/linalg/eig.cuh>
+#include <raft/linalg/gemm.cuh>
+#include <raft/linalg/subtract.cuh>
+#include <raft/matrix/math.cuh>
+#include <raft/matrix/matrix.cuh>
+#include <tsvd/tsvd.cuh>
+#include <common/device_utils.cuh>
+#include <rmm/device_vector.hpp>
+#include <raft/mr/device/allocator.hpp>
+#include <raft/mr/host/allocator.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace ML {
+
+using namespace MLCommon;
+
+
+/**
+ * @brief perform fit operation for kernel PCA. Generates eigenvalues and alphas (unscaled eigenvectors, per sklearn API)
+ * @param[in] handle: cuml handle object
+ * @param[in] input: data to fit using kernel PCA. Size n_rows x n_cols. The size of the data is indicated in prms.
+ * @param[out] alphas: unscaled eigenvectors of the kernel matrix.  scaling occurs in the transform function. Size n_cols * n_components.
+ * @param[out] lambdas: lambdas (eigenvalues) of the principal components. Size n_components * 1.
+ * @param[in] prms: data structure that includes all the parameters from data size to algorithm.
+ * @param[in] stream cuda stream
+ */
+template <typename value_t, typename enum_solver = solver>
+void kpcaFit(const raft::handle_t &handle, value_t *input, value_t *alphas,
+             value_t *lambdas, const paramsKPCA &prms, cudaStream_t stream) {
+  auto cublas_handle = handle.get_cublas_handle();
+  auto allocator = handle.get_device_allocator();
+  ASSERT(prms.n_cols > 1,
+         "Parameter n_cols: number of columns cannot be less than two");
+  ASSERT(prms.n_rows > 1,
+         "Parameter n_rows: number of rows cannot be less than two");
+  ASSERT(
+    prms.n_components > 0,
+    "Parameter n_components: number of components cannot be less than one");
+
+  int n_components = prms.n_components;
+  if (n_components > prms.n_rows) n_components = prms.n_rows;
+
+  Matrix::GramMatrixBase<value_t> *kernel =
+    Matrix::KernelFactory<value_t>::create(prms.kernel, cublas_handle);
+  
+  value_t *kernel_mat;
+  raft::allocate(kernel_mat, prms.n_rows * prms.n_rows);
```

Comment:
These are memory leaks as there is no `cudaFree` for any of these pointers. Use `rmm::dervice_uvector` to allocate memory on the stack. (It's an RAII buffer)

Reply:
Ah good catch thank you.  The kernel matrix and centering matrix are both going to required n_rows x n_rows storage.  I'd like to make good use of device_uvector, but how much stack space can I use?  I think it might work well to at least use it for the diagonal vector that initializes the centering matrix.

Reply:
You misunderstand me, it's an RAII buffer of the ilk of std::vector. Under the hood, it allocates memory with a pointer on the heap, but the usage is safe as you create the object such that it's defined by the scope it exists in. Once the scope ends, it destructs and safely deallocates the memory for you

Reply:
Wow wait that looks potentially amazing.  So you are indeed suggesting I use rmm::device_uvector instead of raft::allocate for the large matrices?  I'm not very familiar with RAII.  Does device_uvector enable one to work with large data on-device, and have it automatically freed upon leaving scope?  Also, I'm struggling to find the meaning of "rmm" and "mr" in RAFT.  Any context on what these stand for?  Thank you!

Reply:
Use rmm::device_uvector for *any* allocations. The definition is in github.com/rapidsai/rmm. CUDA device context has many different ways to allocate memory, which a user can pick and choose from in the Python layer itself, and the underlying allocator is defined with the help of RMM.

Reply:
Wow!  That is amazing.  Everything makes way more sense now.  I had not seen the RMM repository at all.  This resolves several things that I've been confused about, like how .data() calls are supported.  Thank you very much.